### PR TITLE
lxd/lxd: IP filtering on un-managed bridges

### LINKED ISSUE #7477: IP filtering on un-managed bridges
# Required information

 * Distribution: Ubuntu
 * Distribution version: 20.04
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.4
   * LXC version: 4.0.1
   * LXD version: 4.0.1
   * Storage backend in use: dir

# Issue description

LXD does not currently permit the following settings to be set on container network devices which are attached to unmanaged bridges.

I’m trying to achieve the following configuration:

* Bridged networking on the lxd host.
* External router (the lxd host should not NAT this network itself - the external router could be either NATed, or non-NATed - this varies between containers and NICs).
* IP addresses (and gateways, and DNS etc.) statically assigned to the lxd host.
* IP addresses (etc. as above) **either** configured as static IP addresses within each container, or via DHCP (either the lxd host, or an external DHCP server could provide this to the containers).
* Filtering configured on the lxd host, so that it’s not possible for root users within containers to reconfigure the container’s network to successfully spoof the IP address or mac address of another device attached to the bridge network.

This works and is straightforward to implement in all respects, **except** the IP spoofing prevention.

Currently if I try to set this up with an unmanaged bridge (managed by the usual OS means - e.g. netplan) I can setup the following configuration:

`lxc config device add c1 eth1 nic name=eth1 nictype=bridged parent=br1`
`lxc config device set c1 eth1 ipv4.address 10.42.42.42`
`lxc config device set c1 eth1 security.mac_filtering=true`

but extend it like this:

`lxc config device set c1 eth1 security.ipv4_filtering=true`
`lxc config device set c1 eth1 security.ipv6_filtering=true`

… so I can’t guard against IP address spoofing (arp spoofing for ipv4 or NDP spoofing for ipv6) by an attacker which gains control of a container, unless I set up some mechanism external to LXD to accomplish this (e.g. trying to discover the container's corresponding veth devices on the LXD host, and then applying filtering rules there)  - which just feels messy (especially since LXD already supports this for managed bridges).

I may be missing something, but I can't see an obvious reason that this would not be possible to implement for unmanaged bridges in a similar way to the existing fictionality for managed bridges.

See also: https://discuss.linuxcontainers.org/t/ipv4-address-filtering-with-unmanaged-bridges-or-getting-equivalent-functionality/7942/7

### PR DESCRIPTION
**Pull request by my5t3ry**
Maybe resolves #7477. I have the feeling i'm missing out a crutial part, seems to easy. But the code says host and instance filters are set and i wasnt able to spoof the a secured ip address without spoofing the mac address also.

Signed-off-by: Sascha Bast <sash@mischkonsum.org>

### CODE REVIEW

File: shared/util.go [hunk 1]
```diff
@@ -652,6 +652,14 @@ func IsTrue(value string) bool {
 	return StringInSlice(strings.ToLower(value), []string{"true", "1", "yes", "on"})
 }
 
+func IsNotBlank(value string) bool {
```
Comment (tomponline) [inline shared/util.go:655]: @my5t3ry I think we can avoid needing these functions right now. As we use direct `==""` comparison elsewhere for this type of check, such as https://github.com/lxc/lxd/blob/master/lxd/device/nic_bridged.go#L118

I wouldn't want to start introducing inconsistent behaviour.

File: lxd/device/nic_bridged.go [hunk 2]
```diff
@@ -681,8 +681,10 @@ func (d *nicBridged) setFilters() (err error) {
 
 	// If parent bridge is unmanaged check that IP filtering isn't enabled.
 	if err == db.ErrNoSuchObject || n == nil {
-		if shared.IsTrue(d.config["security.ipv4_filtering"]) || shared.IsTrue(d.config["security.ipv6_filtering"]) {
-			return fmt.Errorf("IP filtering requires using a managed parent bridge")
+		if shared.IsTrue(d.config["security.ipv4_filtering"]) && shared.IsBlank(d.config["ipv4.address"]) ||
+			(shared.IsTrue(d.config["security.ipv6_filtering"]) && shared.IsBlank(d.config["ipv6.address"])) {
+			return fmt.Errorf("IP filtering requires using a managed parent bridge." +
+				" If using a unmanaged bridge  accordringly configured address entries are required.")
```
Comment (tomponline) [inline lxd/device/nic_bridged.go:687]: Typo and double space in this line.

File: lxd/device/nic_bridged.go [hunk 3]
```diff
@@ -681,8 +681,10 @@ func (d *nicBridged) setFilters() (err error) {
 
 	// If parent bridge is unmanaged check that IP filtering isn't enabled.
 	if err == db.ErrNoSuchObject || n == nil {
-		if shared.IsTrue(d.config["security.ipv4_filtering"]) || shared.IsTrue(d.config["security.ipv6_filtering"]) {
-			return fmt.Errorf("IP filtering requires using a managed parent bridge")
+		if shared.IsTrue(d.config["security.ipv4_filtering"]) && shared.IsBlank(d.config["ipv4.address"]) ||
+			(shared.IsTrue(d.config["security.ipv6_filtering"]) && shared.IsBlank(d.config["ipv6.address"])) {
+			return fmt.Errorf("IP filtering requires using a managed parent bridge." +
```
Comment (tomponline) [inline lxd/device/nic_bridged.go:686]: We usually just let error messages like this continue on a single line without the need for concatenation as it makes it easier to grep for them.

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
I thought I would take a look at this and see what was going wrong.
I couldn't find anything so I went ahead an added a PR and associated test for it here: https://github.com/lxc/lxd/pull/8886

I think the issues you were experiencing with the tests were environmental.

**Comment by my5t3ry** [comment]
ok, thank you.

**Pull request closed without merging**